lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -1711,27 +1711,36 @@ void GrandCentralPass::runOnOperation() {
   instancePaths = &instancePathCache;
 
   /// Contains the set of modules which are instantiated by the DUT, but not a
-  /// companion or instantiated by a companion.  If no DUT exists, treat the top
-  /// module as if it were the DUT.  This works by doing a depth-first walk of
-  /// the instance graph, starting from the "effective" DUT and stopping the
-  /// search at any modules which are known companions.
+  /// companion, instantiated by a companion, or instantiated under a bind.  If
+  /// no DUT exists, treat the top module as if it were the DUT.  This works by
+  /// doing a depth-first walk of the instance graph, starting from the
+  /// "effective" DUT and stopping the search at any modules which are known
+  /// companions or any instances which are marked "lowerToBind".
   DenseSet<igraph::ModuleOpInterface> dutModules;
   FModuleOp effectiveDUT = dut;
   if (!effectiveDUT)
     effectiveDUT = cast<FModuleOp>(
         *instancePaths->instanceGraph.getTopLevelNode()->getModule());
-  auto dfRange =
-      llvm::depth_first(instancePaths->instanceGraph.lookup(effectiveDUT));
-  for (auto i = dfRange.begin(), e = dfRange.end(); i != e;) {
-    auto module = cast<FModuleLike>(*i->getModule());
-    if (AnnotationSet(module).hasAnnotation(companionAnnoClass)) {
-      i.skipChildren();
-      continue;
+  {
+    SmallVector<igraph::ModuleOpInterface> modules({effectiveDUT});
+    while (!modules.empty()) {
+      auto m = modules.pop_back_val();
+      for (InstanceRecord *a : *instancePaths->instanceGraph.lookup(m)) {
+        auto mod = a->getTarget()->getModule();
+        // Skip modules that we've visited, that are are under the companion
+        // module, or are bound/under a layer block.
+        assert(
+            !a->getInstance()->getParentOfType<LayerBlockOp>() &&
+            "not expecting to see any layerblocks in Grand Central, this will "
```

Comment:
Thanks for the defensive check!!

Your call but if this is reachable and not a logic bug this might be better as a diagnostic and triggering pass failure?

Reply:
It would be reachable from a manually crafted pass pipeline that either doesn't run `LowerLayers` or moves `LowerLayers` after `GrandCentral`. Given that, maybe adding the check and signaling pass failure is better.

Reply:
I'll change this to do the check and signal pass failure.

Reply:
Fixed in https://github.com/llvm/circt/pull/6741/commits/45b7a307dff21f47bbb23dad224b5e362cafcd82. Test added in https://github.com/llvm/circt/pull/6741/commits/0a451ea5b9d85972aefee0565d608d70b952e8aa.